clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1265,30 +1343,56 @@ private void close(Duration timeout, boolean swallowException) {
         }
     }
 
-    /**
-     * Prior to closing the network thread, we need to make sure the following operations happen in the right sequence:
-     * 1. autocommit offsets
-     * 2. release assignment. This is done via a background unsubscribe event that will
-     * trigger the callbacks, clear the assignment on the subscription state and send the leave group request to the broker
-     */
-    private void releaseAssignmentAndLeaveGroup(final Timer timer) {
+    private void autoCommitOnClose(final Timer timer) {
         if (!groupMetadata.get().isPresent())
             return;
 
         if (autoCommitEnabled)
             commitSyncAllConsumed(timer);
 
         applicationEventHandler.add(new CommitOnCloseEvent());
+    }
+
+    private void releaseAssignmentOnClose(final Timer timer) {
+        ConsumerGroupMetadata cgm = groupMetadata.get().orElse(null);
+
+        if (cgm == null)
+            return;
```

Comment:
up to you, but what about:

```suggestion
    if (!groupMetadata.get().isPresent())
        return;
    int memberEpoch = groupMetadata.get().get().generationId();
```

I find it's comes back clearer later on, to refer to member epoch instead of generation id to determine the callback

Reply:
Refactored as suggested 👍